### PR TITLE
#46: Add requirements and design

### LINKED ISSUE #46: Add requirements + design
Add a requirement and design document.

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #46

### CODE REVIEW

File: doc/requirements.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+* `feat~verify-error-code-declarations~1`
+
+Needs: dsn
+
+### Create Error Code Report
+
+`feat~create-error-code-report~1`
+
+**Missing! This feature ios not yet implemented**
```
Comment (pj-spoelders) [inline doc/requirements.md:114]: is*

File: doc/design.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+* `arch~components~1`
+
+Needs: impl, itest
+
+### Duplication Validator
+
+`dsn~duplication-validator~1`
+
+The duplication validator check that each error tag is unique.
```
Comment (tkilias) [inline doc/design.md:69]:
```suggestion
The duplication validator checks that each error tag is unique.
```

Comment (tkilias) [reply doc/design.md:69]: error tag == error code?

File: doc/design.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+The requirements only require that the whole tag is unique. We decided however, to be more strict and also not allow tags that only differ by the number So this implementation considers `E-TEST-1` and `F-TEST-1` as duplicate. This approach has two advantages:
+
+* Users can simply change the type of the code without searching for a new number
```
Comment (tkilias) [inline doc/design.md:73]: I don't understand what is meant here?

**Review by pj-spoelders — APPROVED** [review]
lgtm

**Review by tkilias — CHANGES_REQUESTED** [review]

File: doc/design.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+* A config parser: Component that parses the `error_code_config.yml`
+* A report writer (**missing**): Component that writes the error code report
+
+Needs: dsn
+
+## The Crawler
+
+`dsn~error-declaration-crawler~1`
+
+For implementing the error declaration crawler we decided to use [Spoon](https://spoon.gforge.inria.fr/). This library can create an Abstract Syntax Tree (AST) from Java source code. The error declaration crawler then goes over the AST and searches for invocations of the [error-reporting](https://github.com/exasol/error-reporting-java/) API.
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply doc/design.md:22]: currently, it seems to be supported. If that should change some day, we need to switch ;-)

File: doc/design.md (hunk 3 above)
Comment (jakobbraun) [reply doc/design.md:73]: I tried to clearify

Comment (tkilias) [reply doc/design.md]: ah ok, with users you mean developers, here, maybe use developers instead users here

**Review by tkilias — CHANGES_REQUESTED** [review]
see comment https://github.com/exasol/error-code-crawler-maven-plugin/pull/47#discussion_r649765734

File: doc/design.md (hunk 3 above)
Comment (jakobbraun) [reply doc/design.md]: done

File: doc/design.md [hunk 5]
```diff
@@ -85,7 +85,7 @@ The duplication validator checks that each error identifier is unique.
 The requirements only require that the whole identifier is unique. We decided however, to be more strict and also not allow identifiers that only differ by the number. So this implementation considers `E-TEST-1` and `F-TEST-1` as duplicate. This approach has two advantages:
 
 * Users can simply change the severity of the code without searching for a new number (For example change `E-TEST-1` to `F-TEST-1`)
```
Comment (tkilias) [inline doc/design.md:87]: here users again mean developers?

File: doc/design.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+### Duplication Validator
+
+`dsn~duplication-validator~1`
+
+The duplication validator checks that each error identifier is unique.
+
+The requirements only require that the whole identifier is unique. We decided however, to be more strict and also not allow identifiers that only differ by the number. So this implementation considers `E-TEST-1` and `F-TEST-1` as duplicate. This approach has two advantages:
+
+* Users can simply change the severity of the code without searching for a new number (For example change `E-TEST-1` to `F-TEST-1`)
+* For generating the next free code developers want to keep track of the highest used number. By that limitation they only need to keep track of the highest number per identifier and not per project-shorttag severity combination.
```
Comment (tkilias) [inline doc/design.md:88]:
```suggestion
* For generating the next free code developers only want to keep track of the highest used number. By that limitation they only need to keep track of the highest number per identifier and not per project-shorttag severity combination.
```

**Review by tkilias — CHANGES_REQUESTED** [review]

File: doc/design.md (hunk 5 above)
Comment (jakobbraun) [reply doc/design.md:87]:
```suggestion
* Developers can simply change the severity of the code without searching for a new number (For example change `E-TEST-1` to `F-TEST-1`)
```

**Review by tkilias — APPROVED** [review]

**Pull request merged**